File: src/flights/net/captcha-solver.ts
/**
 * CapSolver / 2captcha-compatible CAPTCHA solver adapter.
 *
 * WHAT: A thin client for the CapSolver REST API (whose request/response shape
 * 2captcha and CapMonster also speak): submit a "create task" describing the
 * captcha challenge, then poll "get task result" until a token is ready. On
 * success it returns the `gRecaptchaResponse` token a login form would otherwise
 * obtain from the embedded reCAPTCHA/Turnstile widget.
 *
 * WHY: Aeroplan and United gate ALL logins behind a Gigya/Auth0 reCAPTCHA
 * (errorCode 401020 "Login Failed Captcha Required"). There is no IP/abuse
 * trigger — it is a server-side decision for every login — so no amount of
 * stealth or proxy rotation clears it. The only programmatic path is to solve
 * the visible reCAPTCHA out-of-band: hand CapSolver the page URL + site key, get
 * back a token, and inject that token into the Gigya `accounts.login` call (or
 * the Auth0 form's `g-recaptcha-response` field). This module produces that
 * token; the login adapters consume it.
 *
 * DORMANT-WITHOUT-KEY: with no CAPSOLVER_API_KEY this never touches the network —
 * `solveCaptcha` logs to STDERR and returns null, so the login adapter falls back
 * to today's "captcha blocked" path and the running daemon is unaffected. The key
 * is read at CALL time (not import time), so flipping it on needs no restart.
 *
 * INJECTABLE HTTP: the actual POST is a `SolverHttp` function. The default uses
 * the global `fetch`; tests pass a fake so the whole create→poll loop runs fully
 * offline with deterministic responses and zero real requests. The poll loop is
 * bounded by a plain integer attempt counter and uses NO wall-clock time and NO
 * randomness — it just increments a counter, so tests are reproducible.
 *
 * Mirrors the dormant-default shape of src/flights/vision-verify.ts and
 * src/flights/healing/parser-autorepair.ts.
 */

/** CapSolver REST base. Both create-task and get-result are POSTs under here. */
const CAPSOLVER_BASE = 'https://api.capsolver.com';

/** Upper bound on how many times we poll getTaskResult before giving up. */
const MAX_POLL_ATTEMPTS = 30;

/** A captcha challenge to solve, expressed in CapSolver/2captcha terms. */
export interface CaptchaTask {
  /** Which widget guards the page. Maps to a CapSolver task `type`. */
  type: 'recaptcha_v2' | 'recaptcha_v3' | 'turnstile';
  /** The URL of the page that hosts the captcha (the login page). */
  websiteUrl: string;
  /** The widget's public site key (reCAPTCHA `data-sitekey` / Turnstile sitekey). */
  websiteKey: string;
  /** reCAPTCHA v3 action name (e.g. "login"); ignored for v2/turnstile. */
  action?: string;
}

/**
 * Injectable HTTP transport: POSTs `body` as JSON to `url` and resolves the
 * parsed JSON response. Kept deliberately minimal so a test fake is trivial and
 * the whole solve flow can run with no real network.
 */
export type SolverHttp = (url: string, body: unknown) => Promise<any>;

/** Map our task type to CapSolver's `task.type` discriminator. */
function capsolverTaskType(t: CaptchaTask['type']): string {
  switch (t) {
    case 'recaptcha_v2':
      return 'ReCaptchaV2TaskProxyLess';
    case 'recaptcha_v3':
      return 'ReCaptchaV3TaskProxyLess';
    case 'turnstile':
      return 'AntiTurnstileTaskProxyLess';
  }
}

function log(msg: string): void {
  console.error(`[captcha-solver ${new Date().toISOString()}] ${msg}`);
}

/**
 * True iff CAPSOLVER_API_KEY is configured (read at call time so the flag can be
 * flipped without restarting the daemon). Login adapters check this to decide
 * whether the captcha path is even available before attempting a solve.
 */
export function isCaptchaSolverConfigured(): boolean {
  return !!process.env.CAPSOLVER_API_KEY && process.env.CAPSOLVER_API_KEY.trim() !== '';
}

/**
 * Default transport: POST JSON via the global `fetch`. Errors (network/non-OK)
 * propagate to `solveCaptcha`, which catches them and returns null — so a
 * transport failure degrades cleanly rather than throwing into a caller.
 */
const defaultHttp: SolverHttp = async (url, body) => {
  const resp = await fetch(url, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });
  if (!resp.ok) {
    throw new Error(`HTTP ${resp.status} from ${url}`);
  }
  return resp.json();
};

/**
 * Solve a captcha challenge, returning the `gRecaptchaResponse`/Turnstile token,
 * or null on dormancy/failure/timeout. NEVER throws.
 *
 * Flow (the CapSolver / 2captcha contract):
 *   1. POST /createTask {clientKey, task:{type, websiteURL, websiteKey, ...}}
 *      → { errorId, taskId } (or { errorId:1, errorDescription } on failure)
 *   2. POST /getTaskResult {clientKey, taskId} repeatedly until
 *      { status:'ready', solution:{ gRecaptchaResponse|token } }
 *      or { status:'failed' } / errorId. Each poll increments a counter; after
 *      MAX_POLL_ATTEMPTS still-processing polls we give up (return null).
 *
 * The returned token is what a login adapter injects into the Gigya
 * `accounts.login({ ..., 'g-recaptcha-response': token })` (or the Auth0 form
 * field) to clear the login captcha.
 *
 * @param task  The challenge to solve (page URL + site key + widget type).
 * @param http  Injectable transport; defaults to `fetch`. Tests pass a fake.
 */
export async function solveCaptcha(
  task: CaptchaTask,
  http: SolverHttp = defaultHttp,
): Promise<string | null> {
  const clientKey = process.env.CAPSOLVER_API_KEY;
  if (!clientKey || clientKey.trim() === '') {
    // Dormant: no key → no network call, caller keeps today's "blocked" path.
    log('CAPSOLVER_API_KEY not set — skipping captcha solve (dormant).');
    return null;
  }

  try {
    // ── 1) Create the task. ──
    const createBody = {
      clientKey,
      task: {
        type: capsolverTaskType(task.type),
        websiteURL: task.websiteUrl,
        websiteKey: task.websiteKey,
        ...(task.type === 'recaptcha_v3' && task.action ? { pageAction: task.action } : {}),
      },
    };
    const created = await http(`${CAPSOLVER_BASE}/createTask`, createBody);
    if (!created || created.errorId) {
      log(
        `createTask failed: ${created?.errorCode ?? 'errorId=' + created?.errorId} ` +
          `${created?.errorDescription ?? ''}`.trim(),
      );
      return null;
    }
    const taskId = created.taskId;
    if (!taskId) {
      log('createTask returned no taskId — aborting.');
      return null;
    }

    // ── 2) Poll for the result, bounded by a plain attempt counter. ──
    // No wall-clock, no randomness: each iteration is one getTaskResult poll,
    // so the loop is fully deterministic and reproducible under test.
    for (let attempt = 0; attempt < MAX_POLL_ATTEMPTS; attempt++) {
      const result = await http(`${CAPSOLVER_BASE}/getTaskResult`, { clientKey, taskId });
      if (!result || result.errorId) {
        log(
          `getTaskResult failed: ${result?.errorCode ?? 'errorId=' + result?.errorId} ` +
            `${result?.errorDescription ?? ''}`.trim(),
        );
        return null;
      }
      if (result.status === 'ready') {
        // CapSolver returns the reCAPTCHA token as `gRecaptchaResponse`;
        // Turnstile returns it as `token`. Accept either.
        const token =
          result.solution?.gRecaptchaResponse ?? result.solution?.token ?? null;
        if (typeof token === 'string' && token.length > 0) {
          log(`solved ${task.type} after ${attempt + 1} poll(s).`);
          return token;
        }
        log('getTaskResult ready but contained no token — treating as failure.');
        return null;
      }
      if (result.status === 'failed') {
        log(`getTaskResult reported status=failed: ${result.errorDescription ?? ''}`);
        return null;
      }
      // status === 'processing' (or anything not-yet-ready): keep polling.
    }

    // Exhausted the attempt budget without a ready token.
    log(`gave up after ${MAX_POLL_ATTEMPTS} polls — token never became ready.`);
    return null;
  } catch (e: any) {
    // Transport or unexpected error: degrade cleanly, never throw into caller.
    log(`solve errored: ${e?.message ?? e}`);
    return null;
  }
}

File: src/flights/scrapers/captcha_solver.py
#!/usr/bin/env python3
"""
CapSolver / 2captcha-compatible captcha solver — Python sibling of
src/flights/net/captcha-solver.ts.

WHY a Python copy: the Aeroplan/United login is a Gigya reCAPTCHA flow that runs
*inside* the Python CDP scraper (united-aeroplan-cdp.py). The captcha appears
mid-flow and the sitekey is only known after the login iframe loads, so the
solve must happen inline in Python — the TS solver can't reach into that moment.
This mirrors the TS contract exactly (createTask -> poll getTaskResult, dormant
without the key, never raises) so behavior stays consistent across languages.

DORMANT-WITHOUT-KEY: solve_captcha() returns None (logging to stderr) when
CAPSOLVER_API_KEY is unset, makes no network call, and NEVER raises — callers
keep today's "captcha blocked" path. The HTTP call is injectable (`http=`) so the
proof test (tests/test-captcha-solver-py.py) runs fully offline with a fake.

NOTE: CapSolver is a PAID third-party service. This module only does anything
when CAPSOLVER_API_KEY is set; setting it implies accepting CapSolver's cost/ToS.
"""

import json
import os
import sys
import time
import urllib.request

# CapSolver REST API base. createTask + getTaskResult are the standard endpoints.
CAPSOLVER_BASE = "https://api.capsolver.com"

# Bound the poll loop by a plain attempt counter (no wall-clock arithmetic), one
# getTaskResult call per attempt with a short sleep between.
MAX_POLL_ATTEMPTS = 30
POLL_INTERVAL_S = 2.0


def _log(msg):
    print(f"[captcha-solver {time.strftime('%H:%M:%S')}] {msg}", file=sys.stderr)


def is_captcha_solver_configured():
    """True iff CAPSOLVER_API_KEY is a non-empty string (read at call time)."""
    key = os.environ.get("CAPSOLVER_API_KEY", "")
    return bool(key and key.strip())


def _default_http(url, body):
    """POST JSON to `url`, return parsed JSON. Used unless an http impl is injected."""
    data = json.dumps(body).encode("utf-8")
    req = urllib.request.Request(
        url, data=data, headers={"Content-Type": "application/json"}, method="POST"
    )
    with urllib.request.urlopen(req, timeout=30) as resp:
        return json.loads(resp.read().decode("utf-8"))


# Map our task types to CapSolver's *ProxyLess task type names.
_TASK_TYPE = {
    "recaptcha_v2": "ReCaptchaV2TaskProxyLess",
    "recaptcha_v3": "ReCaptchaV3TaskProxyLess",
    "turnstile": "AntiTurnstileTaskProxyLess",
}


def solve_captcha(task, http=None):
    """
    Solve a captcha and return the token string, or None on any failure/dormancy.

    task: dict with keys:
      type        - 'recaptcha_v2' | 'recaptcha_v3' | 'turnstile'
      websiteUrl  - the page URL the captcha is on
      websiteKey  - the reCAPTCHA/Turnstile site key
      action      - (optional) reCAPTCHA v3 action

    http: optional callable(url, body) -> dict, injected for offline tests.
          Defaults to a real urllib POST.

    Never raises: dormancy, createTask error, missing taskId, status 'failed',
    missing token, transport error, and poll-budget exhaustion all return None.
    """
    if not is_captcha_solver_configured():
        _log("CAPSOLVER_API_KEY not set — captcha solver dormant, returning None.")
        return None

    api_key = os.environ["CAPSOLVER_API_KEY"].strip()
    do_http = http or _default_http

    task_type = _TASK_TYPE.get(task.get("type"))
    if not task_type:
        _log(f"unsupported captcha task type: {task.get('type')!r}")
        return None

    cap_task = {
        "type": task_type,
        "websiteURL": task.get("websiteUrl"),
        "websiteKey": task.get("websiteKey"),
    }
    if task.get("type") == "recaptcha_v3" and task.get("action"):
        cap_task["pageAction"] = task["action"]

    # 1) createTask
    try:
        created = do_http(
            f"{CAPSOLVER_BASE}/createTask",
            {"clientKey": api_key, "task": cap_task},
        )
    except Exception as e:  # noqa: BLE001 - never propagate transport errors
        _log(f"createTask failed: {e}")
        return None

    if not isinstance(created, dict) or created.get("errorId"):
        _log(f"createTask error: {created.get('errorDescription') if isinstance(created, dict) else created}")
        return None
    task_id = created.get("taskId")
    if not task_id:
        _log("createTask returned no taskId")
        return None

    # 2) poll getTaskResult, bounded by a plain counter
    for attempt in range(1, MAX_POLL_ATTEMPTS + 1):
        try:
            result = do_http(
                f"{CAPSOLVER_BASE}/getTaskResult",
                {"clientKey": api_key, "taskId": task_id},
            )
        except Exception as e:  # noqa: BLE001
            _log(f"getTaskResult failed (attempt {attempt}): {e}")
            return None

        if not isinstance(result, dict) or result.get("errorId"):
            _log(f"getTaskResult error: {result.get('errorDescription') if isinstance(result, dict) else result}")
            return None

        status = result.get("status")
        if status == "ready":
            solution = result.get("solution") or {}
            token = solution.get("gRecaptchaResponse") or solution.get("token")
            if token:
                _log(f"solved {task.get('type')} after {attempt} poll(s).")
                return token
            _log("status ready but no token in solution")
            return None
        if status == "failed":
            _log("solver reported status 'failed'")
            return None
        # status == 'processing' (or unknown): wait and poll again
        time.sleep(POLL_INTERVAL_S)

    _log(f"poll budget exhausted after {MAX_POLL_ATTEMPTS} attempts")
    return None
